tests/derive.rs

```diff
@@ -231,3 +231,42 @@ fn recursive_and_empty_input() {
 
     let _ = Nat5::arbitrary(&mut Unstructured::new(&[]));
 }
+
+#[test]
+fn test_field_attributes() {
```

Comment:
Can we also add some doc tests that use `compile_fail` to check that we return compiler errors when

* `#[arbitrary]` is used multiple times on a single field
* `#[arbitrary(unknown_attr)]` and `#[arbitrary(unknown_attr = unknown_val)]` are used
* `#[arbitrary(value)]` and `#[arbitrary(with)]` are used without RHS values

Should be able to just do something like the following in this file:

```rust
/// Can only use `#[arbitrary]` once per field:
///
/// ```compile_fail
/// use arbitrary::*;
/// #[derive(Arbitrary)]
/// struct Foo {
///     #[arbitrary(with = foo)]
///     #[arbitrary(with = bar)]
///     field: u32,
/// }
/// fn foo(_: &mut Unstructured) -> u32 { todo!() }
/// fn bar(_: &mut Unstructured) -> u32 { todo!() }
/// ```
///
/// Etc...
pub struct CompileFailTests;
```

Reply:
Addressed in https://github.com/rust-fuzz/arbitrary/pull/129/commits/8913ec0d40af472deec4e4ab01eb63ee69d73c80

I was puzzled where to put such doc tests (I don't think users really want to see them in the docs).
Eventually I look into `serde` again, they're using [trybuild](https://github.com/dtolnay/trybuild) to test compile failures. 
So I've decided to introduce `trybuild` to this project too, I think it's a nice tool for this purpose.

Reply:
Was doing doc tests on a `pub struct CompileFailTests` in `tests/derive.rs` not getting picked up by rustdoc?

If not, then we could maybe avoid a new dependency by doing

```rust
// src/lib.rs

#[cfg(all(test, feature = "derive"))]
/// ...
pub struct CompileFailTests {}
```

So that it never shows up in the docs and only is built when we are running tests (with the derive enabled).

Reply:
@fitzgen I just tried doc tests within `derive/src/lib.rs` are not picked up.

> If not, then we could maybe avoid a new dependency by doing

It's `dev-dependency`, shouldn't it be fine? I definitely like `trybuild`, it generates really helpful error messages and allows to structure tests in more easy to understand / maintainable way.

For example, this is what a reported failure looks like:
![image](https://user-images.githubusercontent.com/113512/197013993-cdd9c7f1-9c8e-4c25-a485-9a3613c27062.png)

This way we actually test, the compiler fails in the way we expect it to fail.  Doc tests may give false positives if something in the code gets mistyped.

I would see the cost of using `trybuild` is rather potential failures  if `rustc` changes the way it prints errors.
But still I think pros of `trybuild` outweights its cons.

Please let me know, if you still insist on using doc tests.

Reply:
I think if the test is in `src/lib.rs` rather than `derive/src/lib.rs` it should work? I'd really want to avoid failing tests on nightly but not stable unless as an absolute last resort.